### PR TITLE
WindowServer: Vertically maximize window

### LINKED ISSUE #4123: Vertically maximize window
Under MATE I can vertically maximize a window by middle-clicking on the maximize window icon. I think we should steal this.

### PR DESCRIPTION
**Pull request by Camisul**
Button now can handle middle and right clicks.
Added 2 new handlers in button class: on_right_click for Right mouse
button and on_aux_click for middle mouse button.

Added functionality to vertically maximize window with middle mouse
click on the maximize window button

Closes #4123

### CODE REVIEW

**Comment by nico** [comment]
Since we have this window edge snapping thing already, how about making it so that if you vertical-resize a window and hit the screen edge, it vertically maximizes then? Might be more discoverable than different click behaviors on buttons (which feels kinda early X11-y and weird).

**Comment by Camisul** [comment]
> Since we have this window edge snapping thing already, how about making it so that if you vertical-resize a window and hit the screen edge, it vertically maximizes then? Might be more discoverable than different click behaviors on buttons (which feels kinda early X11-y and weird).

Yea, this sounds like a way better idea, I'll work on it.
Converting to draft for now.

File: Userland/Services/WindowServer/Button.h [hunk 1]
```diff
@@ -52,6 +52,8 @@ class Button : public Weakable<Button> {
     void on_mouse_event(const MouseEvent&);
 
     Function<void(Button&)> on_click;
+    Function<void(Button&)> on_right_click;
+    Function<void(Button&)> on_aux_click;
```
Comment (awesomekling) [inline Userland/Services/WindowServer/Button.h:56]: Let's just call this `on_middle_click` as that names is used everywhere else.

Comment (Camisul) [reply Userland/Services/WindowServer/Button.h]: Done.

File: Userland/Services/WindowServer/Button.cpp [hunk 2]
```diff
@@ -62,22 +62,36 @@ void Button::on_mouse_event(const MouseEvent& event)
...
-    if (event.type() == Event::MouseDown && event.button() == MouseButton::Left) {
+    if (event.type() == Event::MouseDown && event.button() > MouseButton::None && event.button() < MouseButton::Back) {
         m_pressed = true;
         wm.set_cursor_tracking_button(this);
         m_frame.invalidate(m_relative_rect);
         return;
     }
 
-    if (event.type() == Event::MouseUp && event.button() == MouseButton::Left) {
+    if (event.type() == Event::MouseUp && event.button() > MouseButton::None && event.button() < MouseButton::Back) {
```
Comment (awesomekling) [inline Userland/Services/WindowServer/Button.cpp:72]: Same here.

File: Userland/Services/WindowServer/Button.cpp [hunk 3]
```diff
@@ -62,22 +62,36 @@ void Button::on_mouse_event(const MouseEvent& event)
 {
     auto& wm = WindowManager::the();
 
-    if (event.type() == Event::MouseDown && event.button() == MouseButton::Left) {
+    if (event.type() == Event::MouseDown && event.button() > MouseButton::None && event.button() < MouseButton::Back) {
```
Comment (awesomekling) [inline Userland/Services/WindowServer/Button.cpp:65]: This comparison feels really weird and depends on the enum values being in a certain order.
Let's instead test for each of the buttons we want explicitly (i.e Left, Right and Middle)

File: Userland/Services/WindowServer/WindowFrame.cpp [hunk 4]
```diff
@@ -72,6 +72,10 @@ WindowFrame::WindowFrame(Window& window)
         auto button = make<Button>(*this, [this](auto&) {
             WindowManager::the().maximize_windows(m_window, !m_window.is_maximized());
         });
+        button->on_middle_click = [this](auto&) {
+            dbgln("on_middle_click");
```
Comment (awesomekling) [inline Userland/Services/WindowServer/WindowFrame.cpp:76]: Is this meant to be here?

**Review by awesomekling — CHANGES_REQUESTED** [review]

File: Userland/Services/WindowServer/Button.cpp [hunk 5]
```diff
@@ -62,22 +62,36 @@ void Button::on_mouse_event(const MouseEvent& event)
...
+            switch (event.button()) {
+            case MouseButton::Left:
+                if (on_click)
+                    on_click(*this);
+                break;
+
+            case MouseButton::Right:
+                if (on_right_click)
+                    on_right_click(*this);
+                break;
```
Comment (awesomekling) [inline Userland/Services/WindowServer/Button.cpp:88]: This is not used for anything?

File: Userland/Services/WindowServer/Button.cpp [hunk 6]
```diff
@@ -62,22 +62,36 @@ void Button::on_mouse_event(const MouseEvent& event)
...
+
+            default:
+                if (on_middle_click)
+                    on_middle_click(*this);
+                break;
```
Comment (awesomekling) [inline Userland/Services/WindowServer/Button.cpp:93]: The commit message still mentions `on_aux_click`

**Review by awesomekling — CHANGES_REQUESTED** [review]

File: Userland/Services/WindowServer/Button.cpp (hunk 5 above)
Comment (Camisul) [reply Userland/Services/WindowServer/Button.cpp:88]: Yes, I thought adding only `on_middle_click` would be strange

**Pull request merged**